File: server/src/main/java/bgu/spl/net/impl/stomp/Backend/Frames/ExecutableFrame.java
package bgu.spl.net.impl.stomp.Backend.Frames;

import java.util.HashMap;

import bgu.spl.net.genericServers.interfaces.ConnectionHandler;

public abstract class ExecutableFrame extends Frame {

    protected ExecutableFrame(HashMap<String,String> headers, String frameBody, StompCommand command) {
        super(headers, frameBody, command);
    }

    /**
     * Execute the frame according to the command received and return the response
     */
    public abstract String execute(ConnectionHandler<String> handler);

    public static ExecutableFrame parse(String messageToParse){

        Frame frame = Frame.parse(messageToParse);

        if(frame == null)
            return null;
        else if(frame instanceof ExecutableFrame)
            return (ExecutableFrame) frame;
        else
            throw new RuntimeException("Frame is not executable, cannot parse");
    }
}

File: client/src/Frames/ExecutableFrame.cpp
#include "ExecutableFrame.h"

ExecutableFrame::ExecutableFrame(StompCommand command, unordered_map<string, string> headers, string frameBody)
    : Frame(command, headers, frameBody){}

ExecutableFrame *ExecutableFrame::parse(string messageToParse)
{
    Frame* frame = Frame::parse(messageToParse);
    if(frame == nullptr){
        return nullptr;
    }
    else if(ExecutableFrame *executableFrame = dynamic_cast<ExecutableFrame*>(frame)){
        return executableFrame;
    }
    else{
        delete frame;
        throw std::runtime_error("Error: Frame is not executable, cannot parse");
    }
}
